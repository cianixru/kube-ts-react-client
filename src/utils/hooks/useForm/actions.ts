export const SET_FIELD_VALUE = 'SET_FIELD_VALUE';
export const SET_FIELD_TOUCHED = 'SET_FIELD_TOUCHED';
export const SUBMIT_REQUESTED = 'SUBMIT_REQUESTED';
export const SUBMIT_SUCCEEDED = 'SUBMIT_SUCCEEDED';
export const SUBMIT_FAILED = 'SUBMIT_FAILED';
export const SET_ERRORS = 'SET_ERRORS';
