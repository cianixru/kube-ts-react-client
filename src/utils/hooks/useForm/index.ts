// @credits: https://egghead.io/lessons/react-build-your-own-formik-using-react-hooks-with-jared-palmer
import { useReducer, useEffect } from 'react';
import _identity from 'ramda/src/identity';
import {
  SET_FIELD_TOUCHED,
  SET_FIELD_VALUE,
  SET_ALL_FIELD_TOUCHED,
  SET_ERRORS,
} from './actions';

export interface FormAction {
  readonly type: string;
  readonly payload?: any;
}
export type Touched<T> = { readonly [P in keyof T]?: boolean };

export type Values<T> = { readonly [P in keyof T]?: any };

export type Errors<T> = { readonly [P in keyof T]?: string };

export interface State<T> {
  readonly values: Values<T>;
  readonly touched: Touched<T>;
  readonly errors: Errors<T>;
}

export const formReducer = <T>(state: State<T>, action: FormAction) => {
  switch (action.type) {
    case SET_FIELD_VALUE:
      return {
        ...state,
        values: {
          ...state.values,
          [action.payload.name]: action.payload.value,
        },
      };

    case SET_FIELD_TOUCHED:
      return {
        ...state,
        touched: {
          ...state.touched,
          [action.payload.name]: true,
        },
      };

    case SET_ALL_FIELD_TOUCHED:
      return {
        ...state,
        touched: Object.keys(state.values).reduce((acc, next) => {
          return { ...acc, [next]: true };
        }, {}),
      };

    case SET_ERRORS:
      return {
        ...state,
        errors: action.payload,
      };

    default:
      return state;
  }
};

export interface Options<T> {
  readonly initialValues: T;
  readonly onSubmit: (values: Values<T>) => void;
  readonly validate?: (values: Values<T>) => Promise<Errors<T>>;
  readonly validateOnChange?: boolean;
}

export interface FormHook<T> extends State<T> {
  readonly handleBlur: (event: React.ChangeEvent<HTMLInputElement>) => void;
  readonly handleChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  readonly handleSubmit: (event: React.FormEvent<HTMLFormElement>) => void;
}

function useForm<Values>({
  initialValues,
  onSubmit,
  validate = () => Promise.resolve({}),
  validateOnChange = false,
}: Options<Values>): FormHook<Values> {
  const reducerState: State<Values> = {
    errors: {},
    touched: {},
    values: initialValues,
  };

  const [state, dispatch] = useReducer(formReducer, reducerState);

  useEffect(() => {
    if (validateOnChange) {
      validate(state.values)
        .then(() => {
          dispatch({ type: SET_ERRORS, payload: {} });
        })
        .catch(errors => {
          dispatch({ type: SET_ERRORS, payload: errors });
        });
    }
  }, [state.values]);

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    event.persist();

    const { name, value } = event.target;

    dispatch({ type: SET_FIELD_VALUE, payload: { name, value } });
  };

  const handleBlur = (event: React.ChangeEvent<HTMLInputElement>) => {
    event.persist();

    const { name } = event.target;

    dispatch({ type: SET_FIELD_TOUCHED, payload: { name } });
  };

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    dispatch({ type: SET_ALL_FIELD_TOUCHED });

    onSubmit(state.values);
  };

  return {
    handleBlur,
    handleChange,
    handleSubmit,
    ...state,
  };
}

export default useForm;
