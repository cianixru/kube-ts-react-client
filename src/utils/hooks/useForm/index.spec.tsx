import { renderHook, act } from '@testing-library/react-hooks';
import useForm, { formReducer } from './index';
import { TEST_EMAIL, TEST_PASSWORD } from '../../tests/testData';
import {
  SET_FIELD_VALUE,
  SET_FIELD_TOUCHED,
  SET_ERRORS,
  SUBMIT_REQUESTED,
} from './actions';

describe('@formReducer', () => {
  const defaultState = {
    isSubmitting: false,
    values: {
      email: '',
      password: '',
    },
    errors: {},
    touched: {},
  };

  afterEach(() => jest.clearAllMocks());

  it('returns default state', () => {
    const result = formReducer(defaultState, { type: 'NOT_EXISTING_ACTION' });

    expect(result).toEqual(defaultState);
  });

  it('returns changed values for email field', () => {
    const result = formReducer(defaultState, {
      type: SET_FIELD_VALUE,
      payload: { name: 'email', value: TEST_EMAIL },
    });

    expect(result).toEqual({
      ...defaultState,
      values: {
        ...defaultState.values,
        email: TEST_EMAIL,
      },
    });
  });

  it('returns changed touched for email field', () => {
    const result = formReducer(defaultState, {
      type: SET_FIELD_TOUCHED,
      payload: { name: 'email' },
    });

    expect(result).toEqual({
      ...defaultState,
      touched: {
        ...defaultState.touched,
        email: true,
      },
    });
  });

  it('returns changed errors for fields', () => {
    const errors = {
      email: 'invalid email!',
      password: 'password must contains at least 8 characters!',
    };

    const result = formReducer(defaultState, {
      type: SET_ERRORS,
      payload: errors,
    });

    expect(result).toEqual({
      ...defaultState,
      errors,
    });
  });

  it('returns changed touched for email field', () => {
    const result = formReducer(defaultState, {
      type: SUBMIT_REQUESTED,
    });

    expect(result).toEqual({
      ...defaultState,
      isSubmitting: true,
      touched: {
        email: true,
        password: true,
      },
    });
  });
});

describe('@useForm', () => {
  const defaultOptions = {
    initialValues: {
      email: '',
      password: '',
    },
    onSubmit: jest.fn(() => Promise.resolve()),
    validate: jest.fn(() => ({})),
  };

  it('returns initial state', () => {
    const { result } = renderHook(() => useForm(defaultOptions));

    expect(result.current).toMatchSnapshot();
  });

  it('handles change', async () => {
    const { result, waitForNextUpdate } = renderHook(() =>
      useForm(defaultOptions)
    );

    const mock = jest.fn();

    const event = {
      persist: mock as any,
      target: { name: 'email', value: TEST_EMAIL },
    } as React.ChangeEvent<HTMLInputElement>;

    const func: any = async () => {
      result.current.handleChange(event);

      await waitForNextUpdate();
    };

    await act(func);

    expect(result.current).toMatchSnapshot();
    expect(mock).toBeCalledTimes(1);
  });

  it('handles blur', async () => {
    const { result, waitForNextUpdate } = renderHook(() =>
      useForm(defaultOptions)
    );

    const mock = jest.fn();

    const event = {
      persist: mock as any,
      target: { name: 'email' },
    } as React.ChangeEvent<HTMLInputElement>;

    const func: any = async () => {
      result.current.handleBlur(event);

      await waitForNextUpdate();
    };

    await act(func);

    expect(result.current).toMatchSnapshot();
    expect(mock).toBeCalledTimes(1);
  });

  describe('@handleSubmit', () => {
    const values = {
      email: TEST_EMAIL,
      password: TEST_PASSWORD,
    };

    const options = {
      ...defaultOptions,
      initialValues: {
        ...defaultOptions.initialValues,
        ...values,
      },
    };

    const event = {
      preventDefault: () => {},
    } as React.FormEvent<HTMLFormElement>;

    it('when validation and submit successful', async () => {
      const { result, waitForNextUpdate } = renderHook(() =>
        useForm(options)
      );

      const func: any = async () => {
        result.current.handleSubmit(event);

        await waitForNextUpdate();
      };

      await act(func);

      expect(result.current).toMatchSnapshot();
      expect(options.onSubmit).toBeCalledWith(values);
    });

    it('when validation successful and submit unsuccessful', async () => {
      const { result, waitForNextUpdate } = renderHook(() =>
        useForm({
          ...options,
          onSubmit: () => Promise.reject(new Error('submitting error')),
        })
      );

      const func: any = async () => {
        result.current.handleSubmit(event);

        await waitForNextUpdate();
      };

      await act(func);

      expect(result.current).toMatchSnapshot();
      expect(options.onSubmit).toBeCalledWith(values);
    });

    it('when validation unsuccessful', async () => {
      const { result, waitForNextUpdate } = renderHook(() =>
        useForm({
          ...options,
          validate: jest.fn(() => ({
            email: 'Invalid email',
            password: 'Invalid password',
          })),
        })
      );

      const func: any = async () => {
        result.current.handleSubmit(event);

        await waitForNextUpdate();
      };

      await act(func);

      expect(result.current).toMatchSnapshot();
      expect(options.onSubmit).toBeCalledWith(values);
    });
  });
});
