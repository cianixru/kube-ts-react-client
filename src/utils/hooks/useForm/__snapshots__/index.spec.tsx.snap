// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@useForm @handleSubmit when validation and submit successful 1`] = `
Object {
  "errors": Object {},
  "handleBlur": [Function],
  "handleChange": [Function],
  "handleSubmit": [Function],
  "isSubmitting": false,
  "touched": Object {
    "email": true,
    "password": true,
  },
  "values": Object {
    "email": "test@test.com",
    "password": "password",
  },
}
`;

exports[`@useForm @handleSubmit when validation successful and submit unsuccessful 1`] = `
Object {
  "errors": Object {},
  "handleBlur": [Function],
  "handleChange": [Function],
  "handleSubmit": [Function],
  "isSubmitting": false,
  "touched": Object {
    "email": true,
    "password": true,
  },
  "values": Object {
    "email": "test@test.com",
    "password": "password",
  },
}
`;

exports[`@useForm @handleSubmit when validation unsuccessful 1`] = `
Object {
  "errors": Object {
    "email": "Invalid email",
    "password": "Invalid password",
  },
  "handleBlur": [Function],
  "handleChange": [Function],
  "handleSubmit": [Function],
  "isSubmitting": false,
  "touched": Object {
    "email": true,
    "password": true,
  },
  "values": Object {
    "email": "test@test.com",
    "password": "password",
  },
}
`;

exports[`@useForm handles blur 1`] = `
Object {
  "errors": Object {},
  "handleBlur": [Function],
  "handleChange": [Function],
  "handleSubmit": [Function],
  "isSubmitting": false,
  "touched": Object {
    "email": true,
  },
  "values": Object {
    "email": "",
    "password": "",
  },
}
`;

exports[`@useForm handles change 1`] = `
Object {
  "errors": Object {},
  "handleBlur": [Function],
  "handleChange": [Function],
  "handleSubmit": [Function],
  "isSubmitting": false,
  "touched": Object {},
  "values": Object {
    "email": "test@test.com",
    "password": "",
  },
}
`;

exports[`@useForm returns initial state 1`] = `
Object {
  "errors": Object {},
  "handleBlur": [Function],
  "handleChange": [Function],
  "handleSubmit": [Function],
  "isSubmitting": false,
  "touched": Object {},
  "values": Object {
    "email": "",
    "password": "",
  },
}
`;
